buildbot_nix/__init__.py

```diff
@@ -49,95 +73,253 @@ class BuildbotNixError(Exception):
+        else:
+            return (self.skipped_builds_scheduler, props)
 
-    def createTriggerProperties(self, props: Any) -> Any:  # noqa: N802
-        return props
+    @defer.inlineCallbacks
+    def _add_results(self, brid: Any) -> Generator[Any, Any, None]:
+        @defer.inlineCallbacks
+        def _is_buildrequest_complete(brid: Any) -> Generator[Any, Any, bool]:
+            buildrequest: Any = yield self.master.db.buildrequests.getBuildRequest(brid) # TODO: once we bump to 4.1.x use BuildRequestModel | None
+            if buildrequest is None:
+                raise RuntimeError(f"Failed to get build request by its ID")
+            return buildrequest['complete']
+
+        event = ('buildrequests', str(brid), 'complete')
+        yield self.master.mq.waitUntilEvent(event, lambda: _is_buildrequest_complete(brid))
+        builds: Any = yield self.master.db.builds.getBuilds(buildrequestid=brid) # TODO: once we bump to 4.1.x use list[BuildModel]
+        for build in builds:
+            self._result_list.append(build["results"])
+        self.updateSummary()
+
+
+    def prepareSourcestampListForTrigger(self) -> list[dict[str, Any]]: # TODO: ISourceStamp? its defined but never used anywhere an doesn't include `asDict` method
+        ss_for_trigger = {}
+        objs_from_build = self.build.getAllSourceStamps()
+        for ss in objs_from_build:
+            ss_for_trigger[ss.codebase] = ss.asDict()
+
+        trigger_values = [ss_for_trigger[k] for k in sorted(ss_for_trigger.keys())]
+        return trigger_values
+
+    @staticmethod
+    def getDerivationClosure(derivation: str, derivations_inputs: dict[str, set[str]]) -> set[str]:
+        r, size = set([derivation]), 0
+        while len(r) != size:
+            size = len(r)
+            r.update(*[ derivations_inputs[k] for k in r ])
+        return r.difference([derivation])
+
+    @staticmethod
+    def sortJobsByClosures(jobs: list[NixEvalJobSuccess], job_closures: dict[str, set[str]]) -> list[NixEvalJobSuccess]:
+        sorted_jobs = []
+        sorter = graphlib.TopologicalSorter(job_closures)
+
+        for item in sorter.static_order():
+            i = 0
+            while i < len(jobs):
+                if item == jobs[i].drvPath:
+                    sorted_jobs.append(jobs[i])
+                    del jobs[i]
+                else:
+                    i += 1
+
+        return sorted_jobs
+
+    @staticmethod
+    def getFailedDependents(job: NixEvalJobSuccess, jobs: list[NixEvalJobSuccess], job_closures: dict[str, set[str]]) -> list[NixEvalJobSuccess]:
+        jobs = list(jobs)
+        failed_checks: list[NixEvalJob] = [job]
+        failed_paths: list[str] = [job.drvPath]
+        removed = []
+        while True:
+            old_paths = list(failed_paths)
+            for build in list(jobs):
+                deps: set[str] = job_closures.get(build.drvPath, set())
+                print("deps: " + str(deps))
```

Comment:
replace with log function

Reply:
I'm considering replacing with a separate log window in the webui so that the information is easily accessible to users not just the sysadmin.

Reply:
Also works for me.

Reply:
Im planning to grep the codebase for both lix and print 🤣